src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -117,51 +114,40 @@ type Constraint a = SomePolynomial a
 
 -- | Adds a constraint to the arithmetic circuit.
 constraint :: Arithmetic a => Constraint a -> State (ArithmeticCircuit a) ()
-constraint con = modify $ \r -> r { acSystem = insert (con2var con) con (acSystem r) }
+constraint c = zoom #acSystem . modify $ insert (toVar [] c) c
 
 -- | Forces the current variable to be zero.
 forceZero :: forall a . Arithmetic a => State (ArithmeticCircuit a) ()
-forceZero = do
-    r <- get
-    let x   = acOutput r
-        con = polynomial [(one, monomial  (singleton x one))]
-    constraint con
+forceZero = zoom #acOutput get >>= constraint . var
 
 -- | Adds a new variable assignment to the arithmetic circuit.
 -- TODO: forbid reassignment of variables
-assignment :: forall a . (Map Integer a -> a) -> State (ArithmeticCircuit a) ()
-assignment f = modify $ \r -> r { acWitness = (insert (acOutput r) =<< f) . acWitness r }
+assignment :: (Map Integer a -> a) -> State (ArithmeticCircuit a) ()
+assignment f = do
+    i <- insert <$> zoom #acOutput get
+    zoom #acWitness . modify $ (.) (f >>= i)
```

Comment:
As `>>=` here is from `(e ->)` monad, but it is used inside do-notation for `State`, I would once again unfold the usage of `(e ->)` monad in order to not confuse the reader.